File: poses_viewer/src/drill2d/sanityBounds.ts
/**
 * Anatomical sanity bounds — 1:1 mirror of Kotlin SanityBounds. A value outside
 * its band is a tracking glitch: the metric is DROPPED for that frame, never
 * coached on. Bounds are opt-in (unregistered metrics pass through).
 */
import { MetricKey } from './referenceStandard'

const BOUNDS: Record<MetricKey, readonly [number, number]> = {
  elbow_angle: [20, 170],
  shoulder_angle: [5, 175],
  knee_bend: [60, 180],
  torso_lean: [-60, 60],
  shoulder_tilt: [-60, 60],
  hip_flexion: [60, 180], // PROVISIONAL: copied from knee_bend; not yet validated for the hip on footage
}

export function isSane(metricKey: string, value: number): boolean {
  const b = (BOUNDS as Record<string, readonly [number, number]>)[metricKey]
  if (b === undefined) return true
  return value >= b[0] && value <= b[1]
}

File: shared/src/commonMain/kotlin/com/ttcoachai/shared/drill/SanityBounds.kt
package com.ttcoachai.shared.drill

/**
 * Hand-coded anatomical sanity bounds (design decision 2: hand-coded values survive
 * only as sanity bounds). A value outside its band is a tracking glitch — the metric
 * is dropped for that frame, never coached on. Personal baselines, not these bounds,
 * define "correct".
 *
 * Bounds are sourced from [CoreMetricSpecs] (single source of truth); this object's
 * public `isSane` contract is unchanged.
 */
object SanityBounds {

    private val bounds: Map<String, ClosedFloatingPointRange<Double>> =
        CoreMetricSpecs.ALL.mapNotNull { spec -> spec.sanityBounds?.let { spec.key to it } }.toMap()

    /** Metrics without a registered band pass through (bounds are opt-in). */
    fun isSane(metricKey: String, value: Double): Boolean =
        bounds[metricKey]?.contains(value) ?: true
}
